rclpy/rclpy/executors.py

```diff
@@ -0,0 +1,380 @@
+
+    def _take_subscription(self, sub):
+        msg = _rclpy.rclpy_take(sub.subscription_handle, sub.msg_type)
+        return msg
+
+    def _execute_subscription(self, sub, msg):
+        if msg:
+            sub.callback(msg)
+
+    def _take_client(self, client):
+        response = _rclpy.rclpy_take_response(
+            client.client_handle, client.srv_type.Response, client.sequence_number)
+        return response
+
+    def _execute_client(self, client, response):
+        if response:
+            # clients spawn their own thread to wait for a response in the
+            # wait_for_future function. Users can either use this mechanism or monitor
+            # the content of client.response to check if a response has been received
+            client.response = response
+
+    def _take_service(self, srv):
+        request_and_header = _rclpy.rclpy_take_request(
+            srv.service_handle, srv.srv_type.Request)
+        return request_and_header
+
+    def _execute_service(self, srv, request_and_header):
+        if request_and_header is None:
+            return
+        (request, header) = request_and_header
+        if request:
+            response = srv.callback(request, srv.srv_type.Response())
+            srv.send_response(response, header)
+
+    def _make_handler(self, entity, take_from_wait_list, call_callback):
+        """
+        Make a handler that performs work on an entity.
+
+        :param entity: An entity to wait on
+        :param take_from_wait_list: Makes the entity to stop appearing in the wait list
+        :type take_from_wait_list: callable
+        :param call_callback: Does the work the entity is ready for
+        :type call_callback: callable
+        :rtype: callable
+        """
+        gc = self._guard_condition
+        work_cv = self._work_condition
+        num_work = self._num_work_executing
+        is_shutdown = self._is_shutdown
+        # Mark this so it doesn't get added back to the wait list
+        entity._executor_event = True
+
+        def handler():
+            nonlocal entity
+            nonlocal gc
+            nonlocal work_cv
+            nonlocal num_work
+            nonlocal is_shutdown
+            if not entity.callback_group.beginning_execution(entity):
+                # Didn't get the callback, try again later
+                entity._executor_event = False
+                _rclpy.rclpy_trigger_guard_condition(gc)
+                return
+            try:
+                # Track the amount of work being executed
+                if is_shutdown:
+                    return
+                with work_cv:
+                    num_work += 1
+
+                arg = take_from_wait_list(entity)
+
+                # Signal that this has been 'taken' and can be added back to the wait list
+                entity._executor_event = False
+                _rclpy.rclpy_trigger_guard_condition(gc)
+
+                call_callback(entity, arg)
+            finally:
```

Comment:
What kind of exceptions are you trying to catch here? If the lines 190-193 can't raise that they should be before the `try`. If they do, the code might raise before incrementing `num_work` and only decrement the counter in the `finally` block.

Reply:
Good point. The try/finally block was being used to make sure work got decremented if a user's callback raised. I've moved the work incrementing/decrementing to a helper class using __enter__/__exit__ so that the try/finally block could be  narrowed to just the user's callback.

c008cddaef8676af3c1f02b9263d05180864b5f5
223ab33016d7b3764ad87344517a546cc9e4ed25